sphinxdocs/sphinx.bzl

```diff
@@ -0,0 +1,191 @@
+    for format in ctx.attr.formats:
+        output_dir = _run_sphinx(
+            ctx = ctx,
+            format = format,
+            source_path = source_dir_path,
+            output_prefix = "{}/_build".format(ctx.label.name),
+            inputs = inputs,
+        )
+        outputs[format] = output_dir
+    return [
+        DefaultInfo(files = depset(outputs.values())),
+        OutputGroupInfo(**{
+            format: depset([output])
+            for format, output in outputs.items()
+        }),
+    ]
+
+_sphinx_docs = rule(
+    implementation = _sphinx_docs_impl,
+    attrs = {
+        "srcs": attr.label_list(
+            allow_files = True,
+            doc = "Doc source files for Sphinx.",
+        ),
+        "sphinx": attr.label(
+            executable = True,
+            cfg = "exec",
+            mandatory = True,
+            doc = "Sphinx binary to generate documentation.",
+        ),
+        "config": attr.label(
+            allow_single_file = True,
+            mandatory = True,
+            doc = "Config file for Sphinx",
+        ),
+        "formats": attr.string_list(doc = "Output formats for Sphinx to create."),
+        "strip_prefix": attr.string(doc = "Prefix to remove from input file paths."),
+    },
+)
+
+def _create_sphinx_source_tree(ctx):
+    # Sphinx only accepts a single directory to read its doc sources from.
+    # Because plain files and generated files are in different directories,
+    # we need to merge the two into a single directory.
+    source_prefix = ctx.label.name + "/_sources"
+    source_marker = ctx.actions.declare_file(source_prefix + "/__marker")
+    ctx.actions.write(source_marker, "")
+    sphinx_source_dir_path = paths.dirname(source_marker.path)
+    sphinx_source_files = []
+    for orig in ctx.files.srcs:
+        source_rel_path = orig.short_path
+        if source_rel_path.startswith(ctx.attr.strip_prefix):
+            source_rel_path = source_rel_path[len(ctx.attr.strip_prefix):]
+
+        sphinx_source = ctx.actions.declare_file(source_prefix + "/" + source_rel_path)
+        ctx.actions.symlink(
+            output = sphinx_source,
+            target_file = orig,
+            progress_message = "Symlinking Sphinx source %{input} to %{output}",
+        )
+        sphinx_source_files.append(sphinx_source)
+
+    return sphinx_source_dir_path, sphinx_source_files
+
+def _run_sphinx(ctx, format, source_path, inputs, output_prefix):
+    output_dir = ctx.actions.declare_directory("{}/{}".format(output_prefix, format))
+
+    args = ctx.actions.args()
+    args.add("-T")  # Full tracebacks on error
+    args.add("-b", format)
+    args.add("-c", paths.dirname(ctx.file.config.path))
+    args.add("-q")  # Suppress stdout informational text
+    args.add("-j", "auto")  # Build in parallel, if possible
+    args.add("-E")  # Don't try to use cache files. Bazel can't make use of them.
+    args.add("-a")  # Write all files; don't try to detect "changed" files
```

Comment:
nit: it would be great to use parameters with full names (instead of `-T`) we can use something that does not require inline comments. However, since we have inline comments here, it is good.

Reply:
I agree, but as far as I can tell, Sphinx doesn't have long names for them.